app/org/apache/spark/SparkMetricsAggregator.java

```diff
@@ -15,27 +15,69 @@
  */
 
 package org.apache.spark;
+import com.linkedin.drelephant.math.Statistics;
+import org.apache.commons.io.FileUtils;
 
 import com.linkedin.drelephant.analysis.HadoopApplicationData;
 import com.linkedin.drelephant.analysis.HadoopMetricsAggregator;
 import com.linkedin.drelephant.analysis.HadoopAggregatedData;
 import com.linkedin.drelephant.configurations.aggregator.AggregatorConfigurationData;
+import com.linkedin.drelephant.spark.data.SparkApplicationData;
+import com.linkedin.drelephant.spark.data.SparkExecutorData;
+import com.linkedin.drelephant.util.MemoryFormatUtils;
+import java.util.Iterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 public class SparkMetricsAggregator implements HadoopMetricsAggregator {
 
+  private static final Logger logger = LoggerFactory.getLogger(SparkMetricsAggregator.class);
+
   private AggregatorConfigurationData _aggregatorConfigurationData;
 
+  private static final String SPARK_EXECUTOR_MEMORY = "spark.executor.memory";
+
+  private HadoopAggregatedData _hadoopAggregatedData = new HadoopAggregatedData();
+
+
   public SparkMetricsAggregator(AggregatorConfigurationData _aggregatorConfigurationData) {
     this._aggregatorConfigurationData = _aggregatorConfigurationData;
   }
 
   @Override
   public void aggregate(HadoopApplicationData data) {
+    long resourceUsed = 0;
+    long resourceWasted = 0;
+    SparkApplicationData applicationData = (SparkApplicationData) data;
+
+    long perExecutorMem =
+        MemoryFormatUtils.stringToBytes(applicationData.getEnvironmentData().getSparkProperty(SPARK_EXECUTOR_MEMORY, "0"));
+
+    Iterator<String> executorIds = applicationData.getExecutorData().getExecutors().iterator();
+
+    while(executorIds.hasNext()) {
+      String executorId = executorIds.next();
+      SparkExecutorData.ExecutorInfo executorInfo = applicationData.getExecutorData().getExecutorInfo(executorId);
+      // store the resourceUsed in MBSecs
+      resourceUsed += (executorInfo.duration / Statistics.SECOND_IN_MS) * (perExecutorMem / FileUtils.ONE_MB);
+      // maxMem is the maximum available storage memory
+      // memUsed is how much storage memory is used.
+      // any difference is wasted after a buffer of 50% is wasted
+      long excessMemory = (long) (executorInfo.maxMem - (executorInfo.memUsed * 1.5));
```

Comment:
It would be better if we can make the threshold configurable in AggregatorConf.xml.